front/src/modules/trainschedule/components/TimetableV2/hooks.ts

```diff
@@ -0,0 +1,144 @@
+import { useEffect } from 'react';
+
+import dayjs from 'dayjs';
+
+import { enhancedEditoastApi } from 'common/api/enhancedEditoastApi';
+import { useInfraID } from 'common/osrdContext';
+import { isoDateToMs, msToIsoDate } from 'utils/date';
+import { jouleToKwh } from 'utils/physics';
+import { formatKmValue } from 'utils/strings';
+
+import type { TrainScheduleWithDetails } from './types';
+import { extractTagCode, keepTrain } from './utils';
+
+/**
+ * Hook formatting a sorted train schedules array for timetable train cards depending on filtering
+ * @param trainIds the timetable's train ids
+ * @param debouncedFilter filter for train name and labels
+ * @param debouncedRollingstockFilter filter for train's rolling stock metadata
+ * @param validityFilter filter for valid train or not
+ * @param selectedTags filter for train's speed limit tag
+ */
+const useTrainSchedulesDetails = (
+  trainIds: number[],
+  setTrainSchedulesDetails: (trainSchedulesDetails: TrainScheduleWithDetails[]) => void,
+  debouncedFilter: string,
+  debouncedRollingstockFilter: string,
+  validityFilter: string,
+  selectedTags: Set<string | null>
+) => {
+  const infraId = useInfraID();
+
+  const { data: trainSchedulesSummary } =
+    enhancedEditoastApi.endpoints.getV2TrainScheduleSimulationSummary.useQuery(
+      {
+        infra: infraId as number,
+        ids: trainIds,
+      },
+      { skip: !infraId || !trainIds.length }
+    );
+
+  const { data: trainSchedules } = enhancedEditoastApi.endpoints.getV2TrainSchedule.useQuery(
+    {
+      ids: trainIds,
+    },
+    {
+      skip: !trainIds.length,
+    }
+  );
+
+  // We fetch all RS to get the data we need for the advanced filters
+  const { data: { results: rollingStocks } = { results: [] } } =
+    enhancedEditoastApi.endpoints.getLightRollingStock.useQuery({ pageSize: 1000 });
+
+  useEffect(() => {
+    if (trainSchedulesSummary && trainSchedules) {
+      const formatedTrains: TrainScheduleWithDetails[] = trainSchedules
+        .filter((trainSchedule) => {
+          if (!keepTrain(trainSchedule, debouncedFilter)) return false;
+
+          const trainSummary = trainSchedulesSummary[trainSchedule.id];
+
+          // Apply validity filter
+          if (validityFilter === 'valid' && trainSummary.status !== 'success') return false;
+          if (validityFilter === 'invalid' && trainSummary.status === 'success') return false;
```

Comment:
Add some bigger check like this:
```suggestion
        if (validityFilter !== 'both') {
          if (validityFilter === 'valid' && trainSummary.status !== 'success') return false;
          if (validityFilter === 'invalid' && trainSummary.status === 'success') return false;
        }
```

Reply:
Done